src/Stratis.Bitcoin/BlockPulling2/BlockPuller.cs

```diff
@@ -0,0 +1,771 @@
+                }
+
+                await this.AssignDownloadJobsAsync().ConfigureAwait(false);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Loop that continously checks if peers failed to deliver important blocks in given time and penalizes them if they did.</summary>
+        private async Task StallingLoopAsync()
+        {
+            this.logger.LogTrace("()");
+
+            while (!this.cancellationSource.IsCancellationRequested)
+            {
+                try
+                {
+                    await Task.Delay(StallingLoopIntervalMs, this.cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    this.logger.LogTrace("(-)[CANCELLED]");
+                    return;
+                }
+                
+                this.CheckStalling();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private async Task AssignDownloadJobsAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var failedJobs = new List<DownloadJob>();
+            var newAssignments = new Dictionary<uint256, AssignedDownload>();
+
+            lock (this.lockObject)
+            {
+                // First process reassign queue ignoring slots limitations.
+                while (this.reassignedJobsQueue.Count > 0)
+                {
+                    DownloadJob jobToReassign = this.reassignedJobsQueue.Dequeue();
+                    this.logger.LogDebug("Reassigning job {0} with {1} hashes.", jobToReassign.Id, jobToReassign.Hashes.Count);
+
+                    Dictionary<uint256, AssignedDownload> assignments = this.DistributeHashesLocked(ref jobToReassign, ref failedJobs, int.MaxValue);
+
+                    foreach (KeyValuePair<uint256, AssignedDownload> assignment in assignments)
+                        newAssignments.Add(assignment.Key, assignment.Value);
+                }
+
+                // Process regular queue.
+                int emptySlots = this.maxBlocksBeingDownloaded - this.pendingDownloadsCount;
+
+                this.logger.LogDebug("There are {0} empty slots.", emptySlots);
+
+                if (emptySlots > (this.maxBlocksBeingDownloaded / 100) * MinEmptySlotsPercentageToStartProcessingTheQueue)
+                {
+                    while (this.downloadJobsQueue.Count > 0 && emptySlots > 0)
+                    {
+                        DownloadJob jobToAassign = this.downloadJobsQueue.Peek();
+                        this.logger.LogDebug("Reassigning job {0} with {1} hashes.", jobToAassign.Id, jobToAassign.Hashes.Count);
+
+                        Dictionary<uint256, AssignedDownload> assignments = this.DistributeHashesLocked(ref jobToAassign, ref failedJobs, emptySlots);
+                        emptySlots -= assignments.Count;
+                        this.logger.LogDebug("Reassigned {0} hashes.", assignments.Count);
```

Comment:
You have a lot of logs on debug level

Reply:
I do- on purpose. 
Do you think they are placed wrongly and some of them should be trace ones?

Reply:
I know, it just seems too much on that level. Especially when we do not have other levels available because we have this binding with console. So a lot of then I would put on trace level.